awscli/clidocs.py

```diff
@@ -186,18 +190,47 @@ def doc_relateditem(self, help_command, related_item, **kwargs):
         )
         doc.write('\n')
 
-    def _document_enums(self, argument, doc):
+    def _document_enums(self, model, doc):
         """Documents top-level parameter enums"""
-        if hasattr(argument, 'argument_model'):
-            model = argument.argument_model
-            if isinstance(model, StringShape):
-                if model.enum:
-                    doc.style.new_paragraph()
-                    doc.write('Possible values:')
-                    doc.style.start_ul()
-                    for enum in model.enum:
-                        doc.style.li('``%s``' % enum)
-                    doc.style.end_ul()
+        if isinstance(model, StringShape):
+            if model.enum:
+                doc.style.new_paragraph()
+                doc.write('Possible values:')
+                doc.style.start_ul()
+                for enum in model.enum:
+                    doc.style.li('``%s``' % enum)
+                doc.style.end_ul()
+
+    def _document_nested_structure(self, model, doc, stack):
+        """Recursively documents parameters in nested structures"""
+        shape_name = getattr(model, 'name', None)
+        if not shape_name or shape_name in stack:
+            return
+        stack.append(shape_name)
+        if isinstance(model, StructureShape):
```

Comment:
For identifying shape types, let's use the `type_name` attribute as it is a bit more consistently in this module and gives us a bit more freedom if we want to change class names in the future in botocore.